spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/EntityPropertyValueProvider.java

```diff
@@ -51,14 +62,30 @@
 	@SuppressWarnings("unchecked")
 	private <T> T readSingleWithConversion(DatastorePersistentProperty persistentProperty) {
 		Object val = this.entity.getValue(persistentProperty.getFieldName()).get();
-		Class<?> targetType = persistentProperty.getType();
 		if (val == null) {
 			return null;
 		}
-		if (ClassUtils.isAssignable(targetType, val.getClass())) {
-			return (T) val;
+
+		Class<?> targetType = persistentProperty.getType();
+		T result = null;
+
+		//To mirror write conversion, try 1-step conversion first, then try 2-step conversion,
```

Comment:
It's great that you noted here that we specifically want to do the conversions with converters in the exact reverse direction and same intermediate type as the write direction.  I think it would be safer to make some sort of new single class and interface that packages this bi-directional-same-path in a single place so somebody later on doesn't make changes and forget to preserve this property.